---
title: "Lab randomization of samples is important"
author: Miki Bálint, Orsolya Márton, Marlene Schatz, Rolf-Alexander Düring, Hans-Peter Grossart
output:
  pdf_document: default
  html_document: default
---

# Load libraries
``` {r}
library(vegan)
library(effects)
library(lme4)
library(mvabund)
```

```{r eval=FALSE, include=FALSE}
# Other possible packages

# Color palette
# palette(colors())
# library(knitr)
# library(boral)
# library(corrplot)
# library(leaps)
# library(car)
# library(mgcv)
# library(geosphere)
# library(ape) # installed with ctv, infos here: http://www.phytools.org/eqg/Exercise_3.2/
# library(vegan3d)
# library(bvenn)
```

# Read in data
```{r}
# Read abundances
EmblAssign = read.csv(file="../../Data/stechlin_assigned_190915.tab",
                      header=T, sep='|', row.names = 1)

# Read POP and elements data
POP_elem = read.csv(file = "../../Data/stechlin_pop_elements.csv", 
                    header = T, row.names = 1)

# Read experimental setup data
ExpSet = read.csv(file = "../../Data/sample_infos.csv",
                  header = T, row.names = 1)
RepliExp = ExpSet[1:96,]

# re-order the horizon identities
RepliExp$depth.nominal = factor(RepliExp$depth.nominal)

# re-order the nuclear power plant periods
RepliExp$nuclear = factor(RepliExp$nuclear,
                          levels=c("before", 
                                   "during", "after"))
```

How many reads are there?
```{r}
IndexSample = grep("sample.[EMPS]", names(EmblAssign))
sum(EmblAssign[,IndexSample])
```

# Data cleanup
The aim is double: 1) to deal with potential contaminants that may show up in the negative, and 2) to establish a read abundance credibility treshold using the samples that were in the positive controls. 

The assumptions of 1) is that contaminant OTUs observerd in the negatives had their best chances to be present in large numbers only in those negatives since they would have faced PCR competition in any other samples. Thus the most conservative way of dealing with contaminant OTUs is to subtract the maximum number of reads that belongs to them from any other sample where they were also observed.

The reasons behind 2) is that there were only known taxa put into the positives and if others are showing up, they may only come from rare errors. Thus unexpected taxon reads in the positives should give a general sequence rarity treshold for the entire dataset.

### Heads, singletons, internals
This step is specific to the OBITools sequence processing pipeline.

Which columns have the status info for head, internal, singletons?
```{r}
StatusEmbl = EmblAssign[,grepl("obiclean.status", names(EmblAssign))]
```

Keep only sequence variants that were seen at least once as 'head' or 
the 'singleton' count is higher than the 'intermediate' count.

```{r}
EmblHead = EmblAssign[(EmblAssign$obiclean_headcount) > 0 | 
                        EmblAssign$obiclean_singletoncount > EmblAssign$obiclean_internalcount,]
```

### Clean up negative controls
Remove the maximum read number of a sequence variant 
found in a negative control from every sample that 
contains that variant

Select the extraction controls
```{r}
# Extraction controls
ExtCont = grep("sample.EXT", names(EmblHead))

# PCR controls
PCRCont = grep("sample.PCR", names(EmblHead))

# Multiplexing controls
MPXCont = grep("sample.MPX", names(EmblHead))
```

Write out negative control assignments for more analysis
```{r}
NegCont = cbind(name = EmblHead$scientific_name, 
                EmblHead[,c(ExtCont, PCRCont, MPXCont)])

# Aggregate according to taxon
NegRegate = aggregate(. ~ name, NegCont, sum, na.action = na.exclude)
rownames(NegRegate) = NegRegate$name
NegRegate = NegRegate[,2:length(names(NegRegate))]

# Remove taxa that were not seen in negatives
NegRegate = NegRegate[apply(NegRegate,1,sum) > 0,]

# Write for checking negatives
write.csv(file = "negative_control_identities.csv", NegRegate)
```

How many reads are there before cleaning up the negatives?
```{r}
sum(EmblHead[,IndexSample])
```

`sweep` the maximum reads of a sequence variant
in any controls from all samples
```{r warning=FALSE}
# warnings supressed because of a numeric-non-numeric substitution,
# see below
MaxControl = apply(EmblHead[,c(ExtCont,PCRCont,MPXCont)], 1, max)

EmblControlled = EmblHead

# sweep with the MaxControl
EmblControlled[,grep("sample", names(EmblHead))] <- 
  sweep(EmblHead[,grep("sample", names(EmblHead))], 1, MaxControl, "-")

# Set negative values to 0. Warnings are because the non-numeric cells
# There are warnings since many fields are non-numeric.
EmblControlled[EmblControlled < 0] <- 0

# Remove sequence variants with no reads left
EmblControlled = EmblControlled[apply(EmblControlled[,grep("sample", names(EmblHead))],1,sum) > 0,]
```

How many reads are there after cleaning up the negatives?
```{r}
sum(EmblControlled[,IndexSample])
```

Write sequence variants of taxa into table
```{r}
write.csv(file = "variant_sequences.csv", 
          data.frame(name = rownames(EmblControlled),                   
                     name = EmblControlled$scientific_name,
                     genus = EmblControlled$genus_name,
                     family = EmblControlled$family_name,
                     seq = EmblControlled$sequence))
```

### Rare read cleanup with positives controls
Taxa in positive controls

```{r}
PosCont = grep("^sample.POS", names(EmblControlled))

# Abundances in positives and taxonomic annotations
PosAbundances = EmblControlled[,c(PosCont,409,10,9)]

# Sums of all reads / OTU and add to positive table
SumReads = apply(EmblControlled[,IndexSample], 1, sum)
PosAbundances = cbind(PosAbundances, SumReads)

# keep only OTUs that had reads
PosAbundances = PosAbundances[apply(PosAbundances[,1:2],1,sum) > 0,]
```

Write out positive control samples + 
scientific name + genus + family name.
```{r}
write.csv(file="positive_controls.csv", PosAbundances)
```

I manually compared the reads of these OTUs to the 
taxon list and DNA concentration list of the positives.
I looked for a read count rarity threshold of the complete read numbers
`942742` that would reflect the original positive taxon lists.

When an OTU is present with less, than `14` reads in a replicate 
(`0.0015%` of the total reads `sum(ExpSet$reads)` 942742, is considered erroneous and set to `zero`.

```{r warning=FALSE}
RareControlled <- EmblControlled
RareControlled[RareControlled < 14] <- 0
```

How many reads after removing rare observations?
```{r}
sum(RareControlled[,IndexSample])
```


# Effects of laboratory biases
I use all replicates here however weird (similar to negatives or positives, wide away from the other replicates form the same horizon). 

The reason - the aim here is to quantify the effects of potentially systemic lab biases, so all biases qualify except standard contamination issues and rare weird sequences.

## Generate the needed matrices: OTU abundances and corresponding predictors.

```{r}
# Predictors of lab effects
ExpLab = RepliExp[13:96,]

# OTU abundance matrix
OTULab = RareControlled[,IndexSample]
OTULab = data.frame(t(OTULab[,13:96]))

# remove OTUs with zero reads
OTULab = OTULab[,apply(OTULab,2,sum) > 0]

# there is a sample with no reads left, it is removes
OTULab = OTULab[apply(OTULab,1,sum) > 0,]

# Adjust the explaining vars
ExpFilter = rownames(ExpLab) %in% rownames(OTULab)

ExpLab = ExpLab[ExpFilter,]
```

The possible systemic bias to be tested is the person who performed the DNA extraction. The extractions were initially planned to be performed by the first author of the paper, but finally the second half of the extractions were performed by the second author due to an unexpected urgency. Visually it seems that the extractions done by the second author generally yielded more DNA.
```{r}
boxplot(ExpLab$conc ~ ExpLab$person,
     ylab="Extracted DNA concentration (ng/ul)",
     main = "", pch = 19, col="grey", boxwex=0.5, notch = T)
```

Linear mixed effect models are used for the singe response models with the sediment profile identity as the random effect.

## The variables of interest and their predictors 
1. DNA concentration
  + Expected laboratory effect: 
    + sample weight
    + extraction kit
  + Unexpected laboratory effect:
    +lab person
  + Biological effect of interest:
    + the age of the sediment with a higher hump at intermediate ages (visible from an exploratory plot)
    
```{r}
plot(ExpLab$conc ~ ExpLab$age, pch=19,
     xlab = "Age (years)", ylab = "DNA concentration (ng/ul)")
```
    
2. PCR success as indicated by the number of sequence reads per sample. PCR product concentrations were not normalized before multiplexing the samples for sequencing, thus the differences in read numbers may be used to evaluate PCR performance (e.g. `Taberlet papers`)
  + Expected laboratory effect:
    + DNA concentration
    + extraction kit
  + Unexpected laboratory effect:
    + lab person
  + biological effect of interest:
    + sediment age
3. Community structure: diversity indicators (Hill's 1st, 2nd and 3rd numbers) and community composition
  + Expected laboratory effect:
    + read numbers
    + extraction kit
  + Unexpected laboratory effect:
    + lab person
  + biological effect of interest:
    + the effects of the construction/operation period of a nuclear power plant (1960 - 1990). We know that lake communities were strongly changed after building the plant from previous morphology-based works.

## DNA concentration
Fit the full model
```{r}
conc.weight.kit.person.age = 
  lmer(conc ~ weight + kit +
       person + age + I(age^2) + 
         (1|depth.nominal),
     data = ExpLab)

conc.weight.kit.age = 
  lmer(conc ~ weight + kit +
       age + I(age^2) + 
         (1|depth.nominal),
     data = ExpLab)
```

The DNA concentration model is marginally statistically significantly improved by accounting for the lab person.

```{r}
anova(conc.weight.kit.person.age,
      conc.weight.kit.age)
```

Visualize the variable effects. For some reason the `effects` package does not plot into the knitted document, so I use a saved image.
```{r eval=FALSE}
plot(allEffects(conc.weight.kit.person.age,
                multiline=TRUE, confidence.level = 0.95))
```
![Predictors of extracted DNA concentration](conc_effect.png)

Most of the variation in the DNA concentration is explained by the isolation kit. The lab person explains about as much variation as the age of the sediment.
```{r}
anova(conc.weight.kit.person.age)
```

## PCR success
```{r}
read.conc.kit.pers.age = 
  lmer(reads ~ conc + kit + person + age + (1|depth.nominal),
     data = ExpLab)

read.conc.kit.age =
  lmer(reads ~ conc + kit + age + (1|depth.nominal),
       data = ExpLab)
```

The model of PCR success was statistically significantly improved by considering the lab person.
```{r}
anova(read.conc.kit.pers.age,
      read.conc.kit.age)
```

The **DNA concentration** had no consistent effects on the PCR success within a 95% confidence interval (although PCRs produced more reads with higher template concentrations). The **age** of the sediment had no effect on PCR success.

```{r eval=FALSE}
plot(allEffects(read.conc.kit.pers.age,
                multiline=TRUE, confidence.level = 0.95))
```
![Predictors of the obtained read numbers per sample (PCR success)](read_effect.png)

The lab person explained far the most variation in PCR success, with the samples extracted by the second authors yielding consistently less reads (all PCR reactions were performed by the second author). 
```{r}
anova(read.conc.kit.pers.age)
```

# Community structure
## Diversity indicators

Hill's first three numbers of diversity corrspond to richness (Hill's 1), the exponent of Shannon diversity (Hill's 2) and the inverse of the Simpson diversity (Hill's 3).
```{r}
HillLab = renyi(OTULab, scale = c(0,1,2), hill = T)
names(HillLab) = c("hill1", "hill2", "hill3")
```

### Hill's 1 (richness)

```{r}
hill1.read.kit.pers.nucl = 
  lmer(HillLab$hill1 ~ reads + person + kit + nuclear +
         (1|depth.nominal),
     data = ExpLab)

hill1.read.kit.nucl = 
  lmer(HillLab$hill1 ~ reads + kit + nuclear +
         (1|depth.nominal),
     data = ExpLab)
```

Considering the lab person did not 
```{r}
anova(hill1.read.kit.pers.nucl,
      hill1.read.kit.nucl)
```

The richness model was not improved by including the lab person as predictor.

```{r include = FALSE}
plot(allEffects(hill1.read.kit.pers.nucl,
                multiline=TRUE, confidence.level = 0.95))
```
![Predictors of Hill's 1 (richness)](hill1_effect.png)

Most variation in richness was explained by the PCR success, followed by the operation period of the nuclear power plant. The extraction kits also consistently predicted variation in richness, with more OTUs recorded in the replicates extracted with the Macherey-Nagel kit. The variation explained by the lab person was minor. Communities in the lake were less species-rich following the construction of the plant.

```{r}
anova(hill1.read.kit.pers.nucl)
```

### Hill's 2 (exp(Shannon diversity))

```{r}
hill2.read.kit.pers.nucl = 
  lmer(HillLab$hill2 ~ reads + kit + person + nuclear +
         (1|depth.nominal),
     data = ExpLab)

hill2.read.kit.nucl = 
  lmer(HillLab$hill2 ~ reads + kit + nuclear +
         (1|depth.nominal),
     data = ExpLab)
```

The model of Hill's 2 was not improved by considering the lab person's identity.
```{r}
anova(hill2.read.kit.pers.nucl,
      hill2.read.kit.nucl)
```

Hill's 2 was strongly influenced by the PCR success. The replicates extracted with the MN-kit consistently had higher Hill's 2 values. Hill's 2 strongly decreased after the construction of the power plant.

```{r include = FALSE}
plot(allEffects(hill2.read.kit.pers.nucl,
                multiline=TRUE, confidence.level = 0.95))
```
![Predictors of Hill's 2](hill2_effect.png)

The identity of the lab person explains almost no variation in Hill's 2.
```{r}
anova(hill2.read.kit.pers.nucl)
```

### Hill's 3 (inverse Simpson)
Fit the full model.
```{r}
hill3.read.kit.pers.nucl = 
  lmer(HillLab$hill3 ~ reads + kit + person + nuclear +
         (1|depth.nominal),
     data = ExpLab)

hill3.read.kit.nucl = 
  lmer(HillLab$hill3 ~ reads + kit + nuclear +
         (1|depth.nominal),
     data = ExpLab)
```

The lab person identity did not improve the Hill's 2 model.
```{r}
anova(hill3.read.kit.pers.nucl,
      hill3.read.kit.nucl)
```

The effects of the nuclear power plant explained most variation in Hill's 3, resulting in lower diversity after. These were followed by the PCR success and the effects.
```{r include = FALSE}
plot(allEffects(hill3.read.kit.pers.nucl,
                multiline=TRUE, confidence.level = 0.95))
```
![Precictors of Hill's 3](hill3_effect.png)

The lab person explained little variation in Hill's 3.
```{r}
anova(hill3.read.kit.pers.nucl)
```

## Community composition
### Quick NMDS
```{r}
MDS.all <- metaMDS(OTULab)
MDS.all
```

Plot and color the ordination
```{r}
# colors
colfunc <- colorRampPalette(c("green", "brown"))
MyColors = colfunc(83)

par(mar=c(4,4,1,1))
plot(MDS.all$points, type="n",
     xlab="Community axis 1", ylab="Community axis 2")
ordispider(MDS.all, ExpLab$depth.nominal, col="grey")
# plot the points re-ordered according to depth
points(MDS.all$points[order(ExpLab$depth),], pch=19, col = MyColors)
ordisurf(MDS.all, ExpLab$age, add=T, col =
           "grey")
ordiellipse(MDS.all, ExpLab$nuclear,cex=.5, 
            draw="polygon", col=c("green"),
            alpha=100,kind="se",conf=0.95, 
            show.groups=(c("before")))
ordiellipse(MDS.all, ExpLab$nuclear,cex=.5, 
            draw="polygon", col=c("orange"),
            alpha=100,kind="se",conf=0.95, 
            show.groups=(c("during")))
ordiellipse(MDS.all, ExpLab$nuclear,cex=.5, 
            draw="polygon", col=c("blue"),
            alpha=100,kind="se",conf=0.95, 
            show.groups=(c("after")))
# # mylegend = legend(-1, 0.95, c("North, natural","North, heated","South"), 
# ordisurf(MDS.all, ExpLab$DDE.4.4, add=T, col =
#            "red")
```
The replicates are colored according to the sample. Note the weird replicates those all pull toward the negative - positive controls on the complete ordination (not yet shown). The ellipses show the 95% CI group centroids of replicates from before 1960 (green), between 1960-1990 (orange), after 1990 (blue). The countours mar the sediment age. 

### Models of community
I need a general, community-level statistic. `mvabund` cannot deal with random effects: I ommit the horizon identity completely and treat the replicates as independent although they are not

I will compare this with `PERMANOVA` in the `adonis` with `strata` specified as horizons to consider nestedness.

Fit the community model
```{r}
# input data
OTU.mva = mvabund(OTULab)

# full model fit
OTU.read.kit.pers.nucl =
  manyglm(OTU.mva ~ reads + kit + person + nuclear,
          data = ExpLab, family = "negative.binomial")
```

Variation partitioning and predictor stat. significance. Save the ANOVA results.
```{r eval=FALSE}
OTU.anova = anova(OTU.read.kit.pers.nucl,
                  nBoot = 100, test = "LR")

save(OTU.anova, file="lab-methods_OTU_anova.RData")
```

Load the saved ANOVA results
```{r}
load("lab-methods_OTU_anova.RData")
```

Variations explained and statistical significances. Most variation is explained by the biological effect, followed by the person (although both are only marginally significant). The kit doesn't seem to influence the community composition.
```{r}
OTU.anova
```

### Ordination with latent variables
Are there OTUs with weird overdispersion parameters? These are calculated by `mvabund`.

```{r}
hist(OTU.read.kit.pers.nucl$theta, 
     col="grey", xlab="Overdispersion parameters of OTUs", nclass=20)
```

Set the dispersion prior (the last of hypparams) according to the range of **theta** from the `mvabund` run.
```{r}
set.prior = list(type = c("normal","normal","normal","uniform"),
                 hypparams = c(100, 20, 100, 30))
```

Fit the `boral` model only with the OTUs that are not exceedingly overdispersed. Run on the `malloy` since it takes a long time.
```{r eval=FALSE}
# LV ordination done on all OTUs with relatively low overdispersion
save.image(file="Lab_Boral_Data.RData")

LabLVMOrd = boral(OTULab[,OTU.read.kit.pers.nucl$theta < 31],
                  X = ExpLab$reads,
                  family = "negative.binomial", 
                  prior.control = set.prior, 
                  num.lv = 2, n.burnin = 1000, 
                  n.iteration = 10000, n.thin = 9)

save(LabLVMOrd, file="Lab_LV_model_10000-iter.RData")
```

Plot the ordination. Currently just include a previous ordination. The `.Rdata` file with the complete ordination run should be loaded once ready
with `load.image()`, or knitted with `rmarkdown::render("your_doc.Rmd")`.
```{r eval=FALSE}
# Colors and symbols for samples
RepliExpColors = data.frame(RepliExp, 
                            symbols = c(c(rep(22,4), rep(23,2), 
                                          rep(24,4), rep(25,2)),
                                        na.omit(RepliExp$depth.nominal)))
sample_colors = c(levels(factor(RepliExpColors$symbols*10)))
sample_legend = c(levels(factor(RepliExpColors$depth)), 
                  c("Extraction control",
                    "Multiplex control",
                    "PCR control",
                    "Mock community"))

# The LVM ordination plot
par(mfrow = c(1,1), mar = c(4,4.5,1,1))
ordicomm = ordiplot(comm.ord$lv.median, choices = c(1,2), 
                    type = "none", cex =0.5,
                    display = "sites", xlab = "Community axis 1",
                    ylab = "Community axis 2", cex.lab = 1.5)
points(ordicomm, "sites", lwd=2, 
       col=RepliExpColors$symbols*10, 
       pch = RepliExpColors$symbols)
ordispider(ordicomm, RepliExpColors$symbols, 
           col=sample_colors, lwd = 2)
ordisurf(ordicomm, RepliExpColors$depth, add=T, col = "darkgrey")
legend(-1.8, 0.9, sample_legend, border="white", bty="n", lwd = 2,
col=sample_colors, pch = as.numeric(sample_colors)/10, cex = 0.5)
```

# Randomization in the context of experiments/observation VS DNA extraction, PCR setup and sequencing strategy

On Web of Science 14,023 entries were found on Dec. 14, 2016 with the terms "molecul\* AND ecol\*". We ordered the hits from the newest to the oldest and randomly selected 100 papers to evaluate the mentions of **random\* **.
```{r}
# Generate 100 random numbers between 1 - 17,551
write.csv(file="random_numbers_literature.csv", sample(1:14023, 100))
```

There were *XXX* papers that were not suitable for the comparison since they were reviews, or we could not obtain with our available institutional access. We replaced these with a new random set of papers.

```{r}
NotAvailable = c(,)
Reviews = c(,)
```